src/harness/reference_models/pre_iap_filtering/fss_purge.py

```diff
@@ -0,0 +1,296 @@
+from reference_models.propagation import wf_itm
+import pre_iap_util
+# Definition of constants and data types for FSS purge model.
+# Reference Channel bandwidth for FSS OOBE calculation
+REF_BW = 5.e6
+
+# OOBE Margin
+FSS_OOBE_MARGIN = 1
+
+def getFssInfo(fss_record):
+  """ Create a FssProtectionPoint tuple from the fss_record.
+  Args:
+    fss_record: A single FSS record dictionary.
+  Returns:
+    FssProtectionPoint: Tuple holding details about FSS needed for interference calculation.
+  """
+
+  # Extract fields from FSS record. 
+  fss_latitude = fss_record['record']['deploymentParam'][0]['installationParam']['latitude']
+  fss_longitude = fss_record['record']['deploymentParam'][0]['installationParam']['longitude']
+  fss_height = fss_record['record']['deploymentParam'][0]['installationParam']['height']
+  fss_height_type = fss_record['record']['deploymentParam'][0]['installationParam']['heightType']
+  fss_azimuth = fss_record['record']['deploymentParam'][0]['installationParam']['antennaAzimuth']
+  fss_pointing_elevation = fss_record['record']['deploymentParam'][0]['installationParam']['elevation']
+  fss_max_gain_dbi = fss_record['record']['deploymentParam'][0]['installationParam']['antennaGain']
+
+  # Convert the height to AGL if it was AMSL type.
+  if fss_height_type == 'AMSL':
+     fss_height_agl = convertAmslToAgl(fss_latitude, fss_longitude, fss_height)
+  elif fss_height_type == 'AGL':
+     fss_height_agl = fss_height
+  else:
+     raise Exception('Invalid FSS height type.')
+
+  # Populate FssProtectionPoint tuple.abs
+  fss_entity = interference.FssProtectionPoint(latitude=fss_latitude,
+                                               longitude=fss_longitude,
+                                               height_agl=fss_height_agl,
+                                               max_gain_dbi=fss_max_gain_dbi,
+                                               pointing_azimuth=fss_azimuth,
+                                               pointing_elevation=fss_pointing_elevation)
+  return fss_entity
+
+def performPurge(cbsds, fss_entity):
+  """ Implementation of algorithm described in R2-SGN-29.
+  The list of grants to be purged based on the interference caused by their
+  out of band emission are identified and the grants are removed from the CBSDs.
+
+  Args:
+    cbsds: List of CbsdData objects that are in the neighborhood to an FSS.
+    fss_entity: The FssProtectionPoint tuple of the FSS for which the purge is to be performed.
+  """
+
+  while True:
+    # Calculate the OOBE interference value for each CBSD based on the grant with the highest frequency.
+    calculateOobeInterference(cbsds, fss_entity)
+    
+    # Sort the CBSDs based on the calculated interference value.
+    sorted_cbsds = sorted(cbsds, key=lambda cbsd: cbsd['oobe_interference'])
+    # Calculate the threshold value in dBm for the OOBE interference
+    oobe_threshold_value = -129 + 10 * (np.log10(REF_BW/interference.ONE_MHZ) - FSS_OOBE_MARGIN)
```

Comment:
oobe_threshold_value is a constant, so can be moved out of the while loop such that we don't calculate it always. It could even be set once and be used for all FSS protection points of interest

Reply:
As   suggested, moved "oobe_threshold_value" calculation outside while   loop